### PR TITLE
Avoid sending events after disposable have been disposed.

### PR DESCRIPTION
**Pull request by StatusReport**
The API of `RACPassthroughSubscriber` mentions that no more events will be sent after the disposable have been disposed. However, this is not correct if the disposable is disposed on a different thread than the thread that sends the events, as race conditions are possible inside the `send[Next|Completed|Error]` methods, after the `disposable.disposed` test.

### CODE REVIEW

**Comment by StatusReport** [comment]
Notes:
- This is a _huge_ bug in RAC 2.x, IMHO. I can give many examples of bad effects, but for example think of a `UICollectionViewCell` with an image signal that is being replaced upon cell reuse.
- The solution adds a `@synchronized` block around each event that the passthrough subscriber sends. This adds about ~8%-10% performance penalty in my tests. Ideas on how to make it more efficient are welcomed.
- I tried writing test for this, but I'm not sure if it's possible to sanely test this since I need to check if a disposable has been disposed, and the `BOOL` flag in `RACCompoundDisposable` that indicates that is set *before* the inner disposables are being disposed. So without a lot of tricks that may effect the results of the tests I don't see how's that possible.

File: ReactiveCocoa/Objective-C/RACPassthroughSubscriber.m [hunk 1]
```diff
@@ -75,7 +84,9 @@ - (void)sendNext:(id)value {
 		RACSIGNAL_NEXT(cleanedSignalDescription(self.signal), cleanedDTraceString(self.innerSubscriber.description), cleanedDTraceString([value description]));
 	}
```
Comment (mdiep) [inline ReactiveCocoa/Objective-C/RACPassthroughSubscriber.m]: The dtrace macros should probably be moved inside the lock?

And I think this means that the check of `self.disposable.disposed` can be removed. It should probably be replaced with an early exited that checks `if (self.innerSubscriber == nil) return;`.

**Comment by mdiep** [comment]
It doesn't seem like this should matter. When the inner subscriber is disposed, it will stop taking its actions. So even though it doesn't happen in `RACPassthroughSubscriber`, events on the inner subscriber won't be sent after it's disposed.

How did this relate to #2781?

**Comment by StatusReport** [comment]
It does matter, since `RACSubscriber` allows multiple subscriptions over the same `next`, `error` and `complete` blocks. These blocks are `nil`d out when `RACSubscriber`'s disposable is disposed, but `-dispose` is only called when `error` or `complete` is sent, and inside `-dealloc`. Note that it is _not_ called when the disposable given to `-didSubscribeWithDiposable:` is disposed.

So in case we disposed a subscription disposable, the blocks will not be `nil`d out, and additional calls to `next`, `error` or `complete` will still go through to the subscriber until `RACSubscriber` is deallocated. Indeed, the retain cycle between the subscriber's disposable and the subscription disposable is not there after the subscription disposable has been disposed, but if it entered an autoreleasepool, the deallocation of `RACSubscriber` will not happen immediately (and on GCD threads without a runloop [it happens only when they're idle](http://asciiwwdc.com/2015/sessions/718)).

Regarding relation to #2781, I don't want to spoil all the future fun there, but the new implementation of `-switchToLatest` there's a need to dispose the old inner signal and make sure that no additional values from it are being sent (which is exactly the issue I was reporting).

**Comment by mdiep** [comment]
> It does matter, since RACSubscriber allows multiple subscriptions over the same next, error and complete blocks. These blocks are nild out when RACSubscriber's disposable is disposed, but -dispose is only called when error or complete is sent, and inside -dealloc. Note that it is not called when the disposable given to -didSubscribeWithDiposable: is disposed.

That makes sense. Thanks. I'm not very familiar with the RAC 2 internals, so I'm trying to learn as I go here.

My comment above still needs to be addressed.

If you want to try to improve the performance, my understanding is that an `NSLock` will be faster than using `@synchronized`. You could also try using `dispatch_sync` like `RACKVOProxy`; I'm not sure how that compares in performance.

**Comment by tzongw** [comment]
> However, this is not correct if the disposable is disposed on a different thread than the thread that sends the events, as race conditions are possible inside the send[Next|Completed|Error] methods, after the disposable.disposed test.

I think if `dispose` is called on a different thread, whether receiving more events is a chance, since `send` could be called before `dispose` actually taking place and race conditions are still possible.

**Comment by StatusReport** [comment]
@tzongw it is fine for `send*` methods to happen before `dispose`, but after `dispose` has been returned you must not receive additional calls.

@mdiep we're trying to revive #1610, which will replace `RACPassthroughSubscriber` with `RACLiveSubscriber`. It seems that this problem exists there as well and we will fix it before sending it for review. So this PR is on hold for now.

**Comment by mdiep** [comment]
The Objective-C ReactiveCocoa APIs have been moved to https://github.com/ReactiveCocoa/ReactiveObjC/. If you're interested in continuing this PR, please reopen it against that repository. Thanks!

**Pull request closed without merging**